rclcpp/src/rclcpp/context.cpp

```diff
@@ -421,8 +421,8 @@ Context::remove_shutdown_callback(
   const ShutdownCallbackHandle & callback_handle)
 {
   std::mutex * mutex_ptr = nullptr;
-  std::unordered_set<
-    std::shared_ptr<ShutdownCallbackHandle::ShutdownCallbackType>> * callback_list_ptr;
+  std::unordered_set<std::shared_ptr<ShutdownCallbackHandle::ShutdownCallbackType>> *
+    callback_list_ptr = nullptr;
```

Comment:
While I agree that this is a bug, I don't think that this is the right way to fix this.  In particular, if the `shutdown_type` below ends up not being one of the cases, then we'll just go ahead and dereference the nullptr.  I guess that is better in some ways than derferencing random memory, but it is not correct.

I think the proper thing to do here is probably to add a default label to the switch statement below, and probably throw an exception if it is not one of the expected shutdown types.

The same logic goes for below.

Reply:
If only called with values that are handled by the `switch`, this is not a bug. But I agree that a more defensive approach is in order. The way the code is now, introducing actual bugs later is easier than it should be. TBH, I only wanted to silence the warning. I'm glad that you're in favor of a better solution.

The `enum` in the switch statements is only used internally from within `context.cpp`, and when calling `{add,remove,get}_shutdown_callback()`, it is always a constant. So this is actually a compile-time parameter. Adding a default label with a `throw`, i.e., detecting this at runtime, I would not find appropriate. Instead, I made this a compile-time parameter now, so we can `static_assert` that it has one of the supported values.

Let me know what you think!